From e915ce70d4bb7a7ed33d17070f4aaed222f05697 Mon Sep 17 00:00:00 2001
From: "siyong.chen" <siyong.chen@sophgo.com>
Date: Tue, 28 Apr 2026 17:53:22 +0800
Subject: [PATCH 71/84] [vdec] fix jpu dec buf empty process

[description]:NA
[root cause]:NA
[JIRA ID]:CV184XSDK-1224/1220
[chip project]:CV184X
[side effects]:NA
[Affected *.ko/*.so/*.a] :NA

Signed-off-by: siyong.chen <siyong.chen@sophgo.com>
Change-Id: I9199d6cba41798c0ee3e6edc50b479a034ae106e
---
 interdrv/vc_drv/driver/common/vc_proc_comm.c  |  3 +-
 interdrv/vc_drv/driver/dec_api/dec.c          |  2 ++
 interdrv/vc_drv/driver/dec_api/dec_impl.c     |  4 +++
 interdrv/vc_drv/driver/dec_api/hal_dec_task.h |  1 +
 .../vc_drv/driver/dec_api/hal_jpegd_reg.c     | 25 +++++++------
 interdrv/vc_drv/driver/dec_api/jpeg_dec_api.c |  3 +-
 interdrv/vc_drv/driver/dec_api/jpeg_dec_api.h |  2 +-
 interdrv/vc_drv/driver/include/vc_cfg.h       |  1 +
 interdrv/vc_drv/driver/vdec.c                 |  2 +-
 interdrv/vc_drv/driver/vpu_srv/jpu.c          | 35 ++++++++++++++++---
 10 files changed, 60 insertions(+), 18 deletions(-)

diff --git a/interdrv/vc_drv/driver/common/vc_proc_comm.c b/interdrv/vc_drv/driver/common/vc_proc_comm.c
index f6fbb580..094a6bbc 100644
--- a/interdrv/vc_drv/driver/common/vc_proc_comm.c
+++ b/interdrv/vc_drv/driver/common/vc_proc_comm.c
@@ -1595,7 +1595,8 @@ int vdec_proc_show(void *m, void *v)
 				vc_seq_printf(
 					m,
 					"\t Rotation: %u\t PicPoolId: %d\t TmvPoolId: %d\t STATE: %s\n",
-					0, -1, -1, "START");
+					0, p_chn_handle->b_has_vb_pool ? p_chn_handle->vb_pool.h_pic_vb_pool : -1,
+					-1, "START");
 
 				vc_seq_puts(
 					m,
diff --git a/interdrv/vc_drv/driver/dec_api/dec.c b/interdrv/vc_drv/driver/dec_api/dec.c
index 48980a9d..79be3033 100644
--- a/interdrv/vc_drv/driver/dec_api/dec.c
+++ b/interdrv/vc_drv/driver/dec_api/dec.c
@@ -55,6 +55,8 @@ static int jpg_dec_one_pic(void *handle, struct dec_one_pic_cfg_s *stream_cfg, i
         vc_err("Invalid picture configuration\n");
         return VC_ERR_VALUE;
     }
+    osal_memset(&frame_cfg, 0, sizeof(frame_cfg));
+
     ret = vc_dec_oneframe(dec_inst, stream_cfg, &frame_cfg);
     // vc_buffer_invalid_cache(&dec_inst->frame_buf_y);
     // vc_buffer_invalid_cache(&dec_inst->frame_buf_cb);
diff --git a/interdrv/vc_drv/driver/dec_api/dec_impl.c b/interdrv/vc_drv/driver/dec_api/dec_impl.c
index 385e442b..c6a435ee 100644
--- a/interdrv/vc_drv/driver/dec_api/dec_impl.c
+++ b/interdrv/vc_drv/driver/dec_api/dec_impl.c
@@ -152,6 +152,10 @@ enum vc_ret_e vc_dec_oneframe(struct vc_dec_inst_s *dec_impl, struct dec_one_pic
     hal_task->stream_buf = &dec_impl->stream_buf;
     hal_task->bs_len  = stream_cfg->bs_len;
     hal_task->frame_cfg = frame_cfg;
+    if (stream_cfg->vb_phy_addr) {
+       hal_task->frame_cfg->phy_addr[0] = stream_cfg->vb_phy_addr;
+       hal_task->use_vb_pool = 1;
+    }
     ret = vc_dec_hal_gen_regs(hal, hal_task);
     if (ret) {
         vc_err("vc_dec_hal_gen_regs :%-4d failed return %d", __LINE__, ret);
diff --git a/interdrv/vc_drv/driver/dec_api/hal_dec_task.h b/interdrv/vc_drv/driver/dec_api/hal_dec_task.h
index 4abff03b..2d407b51 100644
--- a/interdrv/vc_drv/driver/dec_api/hal_dec_task.h
+++ b/interdrv/vc_drv/driver/dec_api/hal_dec_task.h
@@ -28,6 +28,7 @@ struct hal_dec_task_s {
 	int bs_len;
     unsigned int pts;
     unsigned int dts;
+    unsigned int use_vb_pool;
 
     struct venc_strm_info_s *dec_ret_info;
 };
diff --git a/interdrv/vc_drv/driver/dec_api/hal_jpegd_reg.c b/interdrv/vc_drv/driver/dec_api/hal_jpegd_reg.c
index edccec8f..4b1be91c 100644
--- a/interdrv/vc_drv/driver/dec_api/hal_jpegd_reg.c
+++ b/interdrv/vc_drv/driver/dec_api/hal_jpegd_reg.c
@@ -1415,15 +1415,17 @@ static void setup_frame_info(struct hal_jpegd_ctx_s *ctx, struct hal_dec_task_s
 	unsigned int width_aglin = VC_ALIGN(dec_info->width, 64);
 	unsigned int height_align = VC_ALIGN(dec_info->height, 64);
 
+	if (!task->use_vb_pool) {
 	task->frame_cfg->addr[0] = ctx->frame_buf.ptr;
 	task->frame_cfg->phy_addr[0] = ctx->frame_buf.phy_addr;
+	}
 
 	switch (dec_info->format) {
 	case JPEG_FORMAT_420:
-		task->frame_cfg->addr[1] = ctx->frame_buf.ptr + width_aglin * height_align;
+		task->frame_cfg->addr[1] = task->frame_cfg->addr[0] + width_aglin * height_align;
 		task->frame_cfg->addr[2] = task->frame_cfg->addr[1] + (width_aglin * height_align) / 4;
 
-		task->frame_cfg->phy_addr[1] = ctx->frame_buf.phy_addr + width_aglin * height_align;
+		task->frame_cfg->phy_addr[1] = task->frame_cfg->phy_addr[0] + width_aglin * height_align;
 		task->frame_cfg->phy_addr[2] = task->frame_cfg->phy_addr[1] + (width_aglin * height_align) / 4;
 		task->frame_cfg->length[0] = width_aglin * height_align;
 		if (dec_info->chroma_interleave) {
@@ -1435,10 +1437,10 @@ static void setup_frame_info(struct hal_jpegd_ctx_s *ctx, struct hal_dec_task_s
 		}
 		break;
 	case JPEG_FORMAT_422:
-		task->frame_cfg->addr[1] = ctx->frame_buf.ptr + width_aglin * height_align;
+		task->frame_cfg->addr[1] = task->frame_cfg->addr[0] + width_aglin * height_align;
 		task->frame_cfg->addr[2] = task->frame_cfg->addr[1] + (width_aglin * height_align) / 2;
 
-		task->frame_cfg->phy_addr[1] = ctx->frame_buf.phy_addr + width_aglin * height_align;
+		task->frame_cfg->phy_addr[1] = task->frame_cfg->phy_addr[0] + width_aglin * height_align;
 		task->frame_cfg->phy_addr[2] = task->frame_cfg->phy_addr[1] + (width_aglin * height_align) / 2;
 		task->frame_cfg->length[0] = width_aglin * height_align;
 		if (dec_info->chroma_interleave) {
@@ -1450,10 +1452,10 @@ static void setup_frame_info(struct hal_jpegd_ctx_s *ctx, struct hal_dec_task_s
 		}
 		break;
 	case JPEG_FORMAT_224:
-		task->frame_cfg->addr[1] = ctx->frame_buf.ptr + width_aglin * height_align / 2;
+		task->frame_cfg->addr[1] = task->frame_cfg->addr[0] + width_aglin * height_align / 2;
 		task->frame_cfg->addr[2] = task->frame_cfg->addr[1] + (width_aglin * height_align) / 2;
 
-		task->frame_cfg->phy_addr[1] = ctx->frame_buf.phy_addr + width_aglin * height_align / 2;
+		task->frame_cfg->phy_addr[1] = task->frame_cfg->phy_addr[0] + width_aglin * height_align / 2;
 		task->frame_cfg->phy_addr[2] = task->frame_cfg->phy_addr[1] + (width_aglin * height_align) / 2;
 		if (dec_info->chroma_interleave) {
 			task->frame_cfg->length[0] = width_aglin * height_align / 2;
@@ -1465,20 +1467,20 @@ static void setup_frame_info(struct hal_jpegd_ctx_s *ctx, struct hal_dec_task_s
 		task->frame_cfg->length[2] = width_aglin * height_align;
 		break;
 	case JPEG_FORMAT_444:
-		task->frame_cfg->addr[1] = ctx->frame_buf.ptr + width_aglin * height_align;
+		task->frame_cfg->addr[1] = task->frame_cfg->addr[0] + width_aglin * height_align;
 		task->frame_cfg->addr[2] = task->frame_cfg->addr[1] + width_aglin * height_align;
 
-		task->frame_cfg->phy_addr[1] = ctx->frame_buf.phy_addr + width_aglin * height_align;
+		task->frame_cfg->phy_addr[1] = task->frame_cfg->phy_addr[0] + width_aglin * height_align;
 		task->frame_cfg->phy_addr[2] = task->frame_cfg->phy_addr[1] + width_aglin * height_align;
 		task->frame_cfg->length[0] = width_aglin * height_align;
 		task->frame_cfg->length[1] = width_aglin * height_align;
 		task->frame_cfg->length[2] = width_aglin * height_align;
 		break;
 	case JPEG_FORMAT_400:
-		task->frame_cfg->addr[1] = ctx->frame_buf.ptr;
+		task->frame_cfg->addr[1] = task->frame_cfg->addr[0];
 		task->frame_cfg->addr[2] = task->frame_cfg->addr[1];
 
-		task->frame_cfg->phy_addr[1] = ctx->frame_buf.phy_addr;
+		task->frame_cfg->phy_addr[1] = task->frame_cfg->phy_addr[0];
 		task->frame_cfg->phy_addr[2] = task->frame_cfg->phy_addr[1];
 		task->frame_cfg->length[0] = width_aglin * height_align;
 		task->frame_cfg->length[1] = 0;
@@ -1543,6 +1545,7 @@ enum vc_ret_e hal_jpegd_gen_regs(void *hal, struct hal_dec_task_s *task)
 		return VC_ERR_VALUE;
 	}
 
+	if (!task->use_vb_pool) {
 	if (ctx->frame_buf.phy_addr == 0) {
 		if (vc_buffer_alloc(NULL, &ctx->frame_buf,
 			get_frame_buffer_size(dec_info), "frame_buf")) {
@@ -1564,6 +1567,7 @@ enum vc_ret_e hal_jpegd_gen_regs(void *hal, struct hal_dec_task_s *task)
 			get_frame_buffer_size(dec_info), "frame_buf")) {
 			vc_err("Failed to allocate frm buffer\n");
 			return VC_ERR_MALLOC;
+			}
 		}
 	}
 	setup_frame_info(ctx, task);
@@ -1830,6 +1834,7 @@ enum vc_ret_e hal_jpegd_start(void *hal, struct hal_dec_task_s *task)
 					0x0 + 100, sizeof(VC_JPEG_GBU_C));
 	vc_add_reg_wr_cfg(ctx->dev, &wr_cfg, JPU_CLASS_MC, &reg->vc_jpeg_mc_reg,
 					0x0, sizeof(VC_JPEG_MC_C));
+	if (dec_info->use_hufftab)
 	vc_add_reg_wr_cfg(ctx->dev, &wr_cfg, JPU_CLASS_HUFF, &reg->vc_jpeg_huff_reg,
 					0x0 + 80, (ctx->huff_tab_size + 2) * sizeof(int));
 	vc_add_reg_wr_cfg(ctx->dev, &wr_cfg, JPU_CLASS_QMAT, &reg->vc_jpeg_qmat_reg,
diff --git a/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.c b/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.c
index 717a550c..4dd60fb5 100644
--- a/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.c
+++ b/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.c
@@ -51,13 +51,14 @@ int jpeg_dec_close(drv_jpg_handle jpgHandle)
 	return 0;
 }
 /* send jpu data to decode or encode */
-int jpeg_dec_send_stream(drv_jpg_handle handle, void *data, int length)
+int jpeg_dec_send_stream(drv_jpg_handle handle, void *data, int length, uint64_t vb_phy_addr)
 {
 	struct dec_one_pic_cfg_s stream_cfg, *p_stream_cfg;
 
 	p_stream_cfg = &stream_cfg;
 	p_stream_cfg->bs_addr = data;
 	p_stream_cfg->bs_len = length;
+	p_stream_cfg->vb_phy_addr = vb_phy_addr;
 	dec_api.dev_dec_one_pic(handle, p_stream_cfg, 0);
 	return 0;
 }
diff --git a/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.h b/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.h
index 2825ad26..b6908e93 100644
--- a/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.h
+++ b/interdrv/vc_drv/driver/dec_api/jpeg_dec_api.h
@@ -96,7 +96,7 @@ drv_jpg_handle jpeg_dec_open(drv_jpg_config config);
 /* close and free alloced jpu handle */
 int jpeg_dec_close(drv_jpg_handle jpgHandle);
 /* send jpu data to decode or encode */
-int jpeg_dec_send_stream(drv_jpg_handle handle, void *data, int length);
+int jpeg_dec_send_stream(drv_jpg_handle handle, void *data, int length, uint64_t vb_phy_addr);
 int jpeg_dec_get_frame(drv_jpg_handle handle, void *data, unsigned long int *pu64_hw_time);
 /* release stream buffer */
 int jpeg_dec_release_frame(drv_jpg_handle handle, void *data);
diff --git a/interdrv/vc_drv/driver/include/vc_cfg.h b/interdrv/vc_drv/driver/include/vc_cfg.h
index 98de0086..858b70e7 100644
--- a/interdrv/vc_drv/driver/include/vc_cfg.h
+++ b/interdrv/vc_drv/driver/include/vc_cfg.h
@@ -1210,6 +1210,7 @@ struct dec_one_pic_cfg_s {
 	int nv21;
 	uint64_t pts;
 	uint64_t dts;
+	uint64_t vb_phy_addr;
 };
 struct svc_param_s {
 	/*enable foreground protect according motion info*/
diff --git a/interdrv/vc_drv/driver/vdec.c b/interdrv/vc_drv/driver/vdec.c
index f2f2b71e..dd63ed3f 100644
--- a/interdrv/vc_drv/driver/vdec.c
+++ b/interdrv/vc_drv/driver/vdec.c
@@ -949,7 +949,7 @@ static int jpeg_decode(vdec_chn_context *p_chn_handle, const vdec_stream_s *pst_
 	#endif
 
 	/* send jpeg data for decode or encode operator */
-	ret = jpeg_dec_send_stream(p_chn_handle->p_handle, pst_stream->pu8_addr, pst_stream->u32_len);
+	ret = jpeg_dec_send_stream(p_chn_handle->p_handle, pst_stream->pu8_addr, pst_stream->u32_len, pst_video_frame->video_frame.phyaddr[0]);
 	if (ret != 0) {
 		if ((ret == VDEC_RET_TIMEOUT) && (s32_milli_sec >= 0)) {
 			DRV_VDEC_TRACE("jpeg_dec_send_stream ret timeout\n");
diff --git a/interdrv/vc_drv/driver/vpu_srv/jpu.c b/interdrv/vc_drv/driver/vpu_srv/jpu.c
index 354efa52..dbedb5a0 100644
--- a/interdrv/vc_drv/driver/vpu_srv/jpu.c
+++ b/interdrv/vc_drv/driver/vpu_srv/jpu.c
@@ -664,10 +664,11 @@ static int jpu_run(struct vpu_dev *vpu, struct vpu_task *vpu_task)
 	}
 
 	flush_reg();
-
+	osal_spin_lock(&vpu->dev_spin_lock);
+	vpu->cur_task = vpu_task;
+	osal_spin_unlock(&vpu->dev_spin_lock);
 	vpu_write_relaxed(vpu, REG_JPEG_PIC_START, start_val);
 	/* init current task */
-	vpu->cur_task = vpu_task;
 	vpu_debug_leave();
 	return 0;
 }
@@ -688,18 +689,44 @@ static int jpu_isr(struct vpu_dev *vpu)
 {
 	struct jpu_task *task = NULL;
 	struct vpu_task *vpu_task = vpu->cur_task;
+	int client_type;
 
+	client_type = (enum VPU_CLIENT_TYPE)vpu_task->session->client_type;
 	if (!vpu_task)
 		return OSAL_IRQ_HANDLED;
 	vpu_time_diff(vpu_task, "jpu_done");
 	task = to_jpu_task(vpu_task);
 	task->irq_status = vpu->irq_status;
 	if (task->irq_status & (1 << INT_JPEG_DONE)) {
-		vpu->cur_task = NULL;
 		vpu_write_relaxed(vpu, REG_JPEG_PIC_STATUS, vpu->irq_status);
 		vpu_task_finish(vpu_task->session, vpu_task);
 		return OSAL_IRQ_WAKE_THREAD;
 	}
+	if (client_type == VPU_CLIENT_JPU_DEC) {
+		vpu_log("jpeg irq status: 0x%x\n", task->irq_status);
+		if (task->irq_status & (1 << INT_JPEG_BIT_BUF_EMPTY)) {
+			u32 stream_start_addr;
+			u32 stream_wr_ptr;
+			u32 val;
+
+			stream_start_addr = vpu_read_relaxed(vpu, REG_JPEG_BBC_BAS_ADDR);
+			stream_wr_ptr = vpu_read_relaxed(vpu, REG_JPEG_BBC_WR_PTR);
+			val = (stream_wr_ptr - stream_start_addr) / 256;
+			if((stream_wr_ptr - stream_start_addr) % 256) {
+				val = val + 1;
+			}
+			vpu_write_relaxed(vpu, REG_JPEG_BBC_STRM_CTRL_REG, (1 << 31 | val));
+			vpu_write_relaxed(vpu, REG_JPEG_PIC_STATUS, 1 << INT_JPEG_BIT_BUF_EMPTY);
+			vpu_write_relaxed(vpu, REG_JPEG_INTR_MASK, 0x0);
+			clear_bit(TASK_STATE_HANDLE, &vpu_task->state);
+			task->irq_status = 0;
+			return OSAL_IRQ_HANDLED;
+		}
+		vpu_task_finish(vpu_task->session, vpu_task);
+		jpu_sw_reset(vpu);
+		return OSAL_IRQ_WAKE_THREAD;
+	}
+
 	clear_bit(TASK_STATE_HANDLE, &vpu_task->state);
 	if (task->irq_status & (1 << INT_JPEG_BIT_BUF_FULL)) {
 		u64 stream_buf_start_addr = vpu_read_relaxed(vpu, REG_JPEG_BBC_BAS_ADDR);
@@ -913,4 +940,4 @@ int jpu_remove(struct vpu_dev *vpu)
 	vpu_dev_remove(&enc->vpu);
 	osal_vfree(enc);
 	return 0;
-}
\ No newline at end of file
+}
-- 
2.25.1

